docs/source/extension/linting.py

```diff
@@ -0,0 +1,152 @@
+    # \s+          - 1+ whitespace after #
+    # (.+?)        - Non-greedy capture of heading text (group 1)
+    # \s*          - Optional whitespace after heading text
+    # #*           - Optional closing #s (some markdown flavors allow this)
+    # $            - End of line
+    md_pattern = re.compile(r'^#{2,}\s+(.+?)\s*#*$', re.MULTILINE)
+    headings += md_pattern.findall(content)
+
+    # reST underlined headings pattern explanation:
+    # We need to identify the document structure to exclude top-level headings
+    # This simplified approach looks for specific underline characters
+    # typically used for section levels below the top level
+    #
+    # ^            - Start of line
+    # ([^\n]+)     - Capture heading text (group 1) - any chars except newline
+    # \n           - Newline after heading text
+    # ([-`:"'~+^_#*]) - Capture underline character (group 2) from valid set
+    # \2+          - 1+ repeats of same underline character
+    # \s*$         - Optional trailing whitespace
+    #
+    # Note: Now includes '-' but excludes '=' (which is reserved for top-level)
+    rst_pattern = re.compile(
+        r'^([^\n]+)\n'  # Heading text
+        r'([-`:"\'~+^_#*])'  # First underline char (now includes - but excludes =)
+        r'\2+\s*$',  # Repeat same char + whitespace
+        re.MULTILINE)
+    headings += [m[0] for m in rst_pattern.findall(content)]
+
+    for heading in headings:
+        violations = []
+        # Split on whitespace, preserving punctuation with words
+        words = re.findall(r'\S+', heading)
+        step_pattern = re.compile(r'^step$', re.IGNORECASE)
+        number_pattern = re.compile(
+            r'^\d+[:\-.]$')  # Matches digits followed by : or - or .
+        identifier_pattern = re.compile(
+            r'^[A-Za-z0-9]+$', re.IGNORECASE)  # Alphanumeric identifier
+        punctuation_pattern = re.compile(
+            r'^[:-]$')  # Punctuation as separate word
+
+        for i in range(len(words)):
+            original_word = words[i]
+            # Remove ALL leading/trailing non-alphanumeric characters
+            stripped_word = re.sub(r'^\W+|\W+$', '', original_word)
+
+            # Skip allowed terms and acronyms (check stripped version)
+            if stripped_word in ALLOWED_TERMS or stripped_word.isupper():
+                continue
+
+            # Check for step pattern: "Step [identifier][punctuation?] [word]"
+            if ((i >= 2 and step_pattern.match(words[i - 2]) and
+                 (number_pattern.match(words[i - 1]) or
+                  identifier_pattern.match(words[i - 1]))) or
+                (i >= 3 and step_pattern.match(words[i - 3]) and
+                 identifier_pattern.match(words[i - 2]) and
+                 punctuation_pattern.match(words[i - 1]))):
+                continue
+
+            # Check if previous word is a number with punctuation
+            if i >= 1 and number_pattern.match(words[i - 1]):
+                continue
```

Comment:
I think we can just simplify this to skip whenever the previous word ends in a punctuation. It doesn't need to specifically be `Step 1: Title` or `1. Title`. `First part: Second part` I think should also be okay which will simplify the logic.